src/library/scala/concurrent/Future.scala

```diff
@@ -115,9 +115,7 @@ trait Future[+T] extends Awaitable[T] {
    */
   @deprecated("use `foreach` or `onComplete` instead (keep in mind that they take total rather than partial functions)", "2.12.0")
   def onSuccess[U](pf: PartialFunction[T, U])(implicit executor: ExecutionContext): Unit = onComplete {
-    case Success(v) =>
-      pf.applyOrElse[T, Any](v, Predef.identity[T]) // Exploiting the cached function to avoid MatchError
-    case _ =>
+    t => if (t.isInstanceOf[Success[T]]) pf.applyOrElse[T, Any](t.asInstanceOf[Success[T]].value, Future.id[T])
```

Comment:
Interesting! Is the new version without partial function/pattern matching more JIT-friendly?

Reply:
Yes, pattern-matching produces more bytecode, which makes it less-likely to inline.

Reply:
@fwbrasil I've updated the code to have a callback tree with only 2 nodes.